htlcswitch/switch.go

```diff
@@ -0,0 +1,599 @@
+	// CloseRegular indicates a regular cooperative channel closure
+	// should be attempted.
+	CloseRegular ChannelCloseType = iota
+
+	// CloseForce indicates that the channel should be forcefully closed.
+	// This entails the broadcast of the commitment transaction directly on
+	// chain unilaterally.
+	CloseForce
+
+	// CloseBreach indicates that a channel breach has been dtected, and
+	// the link should immediately be marked as unavailable.
+	CloseBreach
+)
+
+// ChanClose represents a request wto close a particular channel specified by
+// its outpoint.
+type ChanClose struct {
+	CloseType ChannelCloseType
+
+	ChanPoint *wire.OutPoint
+
+	Updates chan *lnrpc.CloseStatusUpdate
+	Err     chan error
+}
+
+// pendingPayment represents the payment which made by user and waits for
+// updates to be sent in channels about has it been rejected or processed
+// successfully.
+type pendingPayment struct {
+	err      chan error
+	preimage chan [sha256.Size]byte
+}
+
+// Switch is a central messaging bus for all incoming/outgoing htlc's.
+// The goal of the switch is forward the incoming/outgoing htlc messages from
+// one channel to another, and also propagate the settle/fail htlc messages
+// back to original requester. In order to better understand the whole view:
+//	alice			   bob				   carol
+//	server - <-connection-> - server - - <-connection-> - - - server
+//	 |		   	  |				    |
+//   alice htlc			bob htlc		       carol htlc
+//     switch			switch			         switch
+//	|			 |       \			   |
+//	|			 |        \			   |
+// alice       <-------->  first bob    second bob   <---------> carol
+// channel link	    	  channel link   channel link		channel link
+type Switch struct {
+	started  int32
+	shutdown int32
+	wg       sync.WaitGroup
+	quit     chan bool
+	mutex    sync.RWMutex
+
+	// pendingPayments is correspondence of payments and its hashes,
+	// which is used to save the payments which made by user.
+	pendingPayments map[[sha256.Size]byte]*pendingPayment
+
+	// circuits is an structure which is used to forward the settle/fail
+	// htlc updates back to the add htlc initiator.
+	circuits *circuitMap
+
+	// links is a map of channel output and channel link which manages
+	// this channel.
+	links map[wire.OutPoint]ChannelLink
+
+	// commands is used to propagate the commands to the handler goroutine.
+	commands chan interface{}
+}
+
+// New creates the new instance of htlc switch.
+func New() *Switch {
+	return &Switch{
+		circuits:        newCircuitMap(),
+		links:           make(map[wire.OutPoint]ChannelLink),
+		pendingPayments: make(map[[sha256.Size]byte]*pendingPayment),
+		commands:        make(chan interface{}),
+		quit:            make(chan bool, 1),
+	}
+}
+
+// SendUpdate is used by other subsystems which aren't belong to htlcswicth
+// package in order to send the payment.
+// TODO(andrew.shvv) lnwire messages should be hidden from upper levels.
+func (s *Switch) SendUpdate(dest routing.HopID, htlc *lnwire.UpdateAddHTLC) (
```

Comment:
But in other hand we may consider switch as something that only takes the packets as input... and forward them, but in this case the send update function should be removed at all, but this will lead to other abstraction inconsistence.... I will change it how it was requested, but I think we need collaborate a bit on abstractions that we have in order to develop unique view on how this should be done in future.